codeBlock-1984/bankr
# #165480040 Secure routes with jwt (JWT)

### What does this PR do?
Implements role-based authentication in the server, securing all routes using JSON Web Token.

### Description of Task to be completed?
- Create authenticator class for authenticator middleware functions
- Build isAuth function for validating request token
- Build permission level functions for authorizing users based on user roles
- Secure routes using authenticator middleware
- Add request tokens to tests to enable them to pass as before

### How should this be manually tested?
1. Follow the instructions on the README file to start the development environment on your local machine.
2. In your terminal run ```curl --request GET http://localhost:3000/api/v1/auth/signup -d 'firstName=John&lastName=Winston&email=johnwinston@gmail.com&password=password&type=client&isAdmin=false' ```
3. You should get a response object with a response status of **201** and a data array of a single object of signup details including an **id** and a **token**.
4. Run the following command ```curl -H "x-auth-token: token" --request POST http://localhost:3000/api/v1/accounts -d 'accountNumber=1029394848&owner=:id&type=savings&openingBalance=20000.40&status=active```. Where **token** and **id** are the from your signup details
5. You should receive a response with status code **201** and newly-created account details.

### What are the relevant pivotal tracker stories?
[#165480040](https://www.pivotaltracker.com/story/show/165480040)
